wiring/inc/spark_wiring_buffer.h

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2024 Particle Industries, Inc.  All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <cstring>
+#include <cstdint>
+
+#include "spark_wiring_string.h"
+#include "spark_wiring_vector.h"
+
+namespace particle {
+
+/**
+ * A dynamically allocated buffer.
+ */
+class Buffer {
+public:
+    /**
+     * Construct an empty buffer.
+     */
+    Buffer() = default;
+
+    /**
+     * Construct a buffer of a specific size.
+     *
+     * The buffer is filled with zero bytes.
+     *
+     * @param size Buffer size.
+     */
+    explicit Buffer(size_t size) :
+            d_(size) {
+    }
+
+    /**
+     * Construct a buffer by copying existing data.
+     *
+     * @param data Data to copy.
+     * @param size Data size.
+     */
+    Buffer(const char* data, size_t size) :
+            d_(data, size) {
+    }
+
+    /**
+     * Get a pointer to the buffer data.
+     *
+     * @return Buffer data.
+     */
+    char* data() {
+        return d_.data();
+    }
```

Comment:
The original `Buffer` implementation that we used internally exposed its data as `char*` so I did the same here to reduce changes in unrelated parts of Device OS that use this class. The Wiring library tends to use `uint8_t*` for binary data though so I'm wondering if it would be better to use that type instead.